include/nil/crypto3/zk/snark/systems/plonk/placeholder/gates_argument.hpp

```diff
@@ -105,34 +107,76 @@ namespace nil {
                             transcript_type& transcript) {
                         PROFILE_PLACEHOLDER_SCOPE("gate_argument_time");
 
-                        std::uint32_t extended_domain_size = original_domain->m * std::pow(2, ceil(std::log2(max_gates_degree)));
+                        std::uint32_t extended_domain_size = original_domain->m * 
+                            std::pow(2, ceil(std::log2(max_gates_degree)));
                         
                         const plonk_polynomial_dfs_table<FieldType, typename ParamsType::arithmetization_params>
                             extended_column_polynomials = resize(column_polynomials, extended_domain_size);
 
                         typename FieldType::value_type theta = transcript.template challenge<FieldType>();
 
-                        std::array<math::polynomial_dfs<typename FieldType::value_type>, argument_size> F;
-                        F[0] = math::polynomial_dfs<typename FieldType::value_type>(0, extended_domain_size, FieldType::value_type::zero());
+                        auto value_type_to_polynomial_dfs = [&assignments=extended_column_polynomials](
+                            const typename variable_type::assignment_type& coeff) {
+                                return polynomial_dfs_type(0, assignments.rows_amount(), coeff);
+                            };
+                        math::expression<polynomial_dfs_variable_type> expr = 
+                            value_type_to_polynomial_dfs(FieldType::value_type::zero());
 
                         typename FieldType::value_type theta_acc = FieldType::value_type::one();
 
+                        // Every constraint has variable type 'variable_type', but we want it to use
+                        // 'polynomial_dfs_variable_type' instead. The only difference is the coefficient type
+                        // inside a term. We want the coefficients to be dfs polynomials here.
+                        math::expression_variable_type_converter<variable_type, polynomial_dfs_variable_type> converter(
+                            value_type_to_polynomial_dfs);
+
                         const auto& gates = constraint_system.gates();
                         for (const auto& gate: gates) {
-                            math::polynomial_dfs<typename FieldType::value_type> gate_result(
-                                0, extended_domain_size, FieldType::value_type::zero());
+                            math::expression<polynomial_dfs_variable_type> gate_result = 

```

Comment:
Here is the main change, everything else was done for this. We want to able to create expressions with dfs polynomials as coefficients.